### DISCUSSION TITLE
일반적으로 웹 사이트의 전역 스코프를 그대로 두고 건드리지 않는 것이 좋은 이유는 무엇인가요?

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#51] [스코프]
- 일반적으로 웹 사이트의 전역 스코프를 그대로 두고 건드리지 않는 것이 좋은 이유는 무엇인가요?
  - 아래 실행 결과를 스코프와 관련 지어 설명해주세요.
     ```js
      (function() {
          var localVar = 'I am local';
          console.log(localVar);
      })();
      console.log(localVar);
      ```

**Comment by publdaze** [comment]
> - 일반적으로 웹 사이트의 전역 스코프에서 직접 코드를 다루는 게 안 좋은 이유는 무엇인가요?
- 전역 스코프는 모든 자바스크립트 파일이 접근 가능
- 파일 별로 서로 식별자가 겹치면 의도하지 않게 데이터가 변경될 수 있다는 단점이 있음
- 따라서 가급적 전역 스코프에서 코드를 작성하지 않고 즉시 실행 함수나 모듈패턴 등을 사용해 전역을 보호하는 것이 좋음

> - 아래 실행 결과를 스코프와 관련 지어 설명해주세요.
>      ```js
>       (function() {
>           var localVar = 'I am local';
>           console.log(localVar);
>       })();
>       console.log(localVar);
>     ```

- 코드가 즉시 실행 함수로 감싸져 있음
- 즉시 실행 함수는 한 번만 호출되며 함수 정의와 동시에 호출됨
- 코드에서 즉시 실행 함수가 호출되었고, 함수 몸체 내부는 지역 스코프에 해당하기 때문에 localVar은 지역 변수가 되어 해당 즉시 실행 함수에서만 유효하다.
- 따라서 첫번째 출력에서는 'I am local'이 되고, 지역 스코프에서 벗어난 2번째 출력에서는 reference 에러가 발생한다.

**Comment by pipisebastian** [comment]
> 일반적으로 웹 사이트의 전역 스코프를 그대로 두고 건드리지 않는 것이 좋은 이유는 무엇인가요?

- 전역스코프는 어디서든 접근이 가능하기 때문에, 의도치 않게 값이 변경될 수 있기 때문입니다.

>아래 실행 결과를 스코프와 관련 지어 설명해주세요.
```tsx
 (function() {
     var localVar = 'I am local';
     console.log(localVar);
 })();
 console.log(localVar);
```

- 3 line에서 찍히는 로그는 'I am local', 5 line에서는 reference error를 반환합니다.
- function은 즉시실행함수로, 즉시실행함수 내부 변수는, 지역변수로 취급됩니다. 이에 localVar는 해당 스코프에서만 접근 가능하기에, 함수 내부에서는 참조가 가능하나, 전역 스코프에서는 참조가 불가능합니다.
